lib/Controller/CardApiController.php

```diff
@@ -0,0 +1,242 @@
+<?php
+/**
+ * @copyright Copyright (c) 2018 Ryan Fletcher <ryan.fletcher@codepassion.ca>
+ *
+ * @author Ryan Fletcher <ryan.fletcher@codepassion.ca>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+ namespace OCA\Deck\Controller;
+
+ use OCP\AppFramework\ApiController;
+ use OCP\AppFramework\Http;
+ use OCP\AppFramework\Http\DataResponse;
+ use OCP\IRequest;
+
+ use OCA\Deck\Controller\Helper\ApiHelper;
+ use OCA\Deck\Service\BoardService;
+ use OCA\Deck\Service\StackService;
+ use OCA\Deck\Service\CardService;
+
+ /**
+ * Class BoardApiController
+ *
+ * @package OCA\Deck\Controller
+ */
+class CardApiController extends ApiController {
+	private $cardService;
+	private $boardService;
+	private $stackService;
+	private $userId;
+	private $apiHelper;
+
+	/**
+	 * @param string $appName
+	 * @param IRequest $request
+	 * @param CardService $service
+	 * @param $userId
+	 */
+	public function __construct($appName, IRequest $request, CardService $cardService, BoardService $boardService, StackService $stackService, $userId) {
+		parent::__construct($appName, $request);
+		$this->boardService = $boardService;
+		$this->cardService = $cardService;
+		$this->stackService = $stackService;
+		$this->userId = $userId;
+		$this->apiHelper = new ApiHelper();
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @CORS
+	 * @NoCSRFRequired
+	 *
+	 * Get a specific card.
+	 */
+	public function get() {
+		$boardError = $this->apiHelper->entityHasError($this->request->params['boardId'], 'board', $this->boardService);
+		if ($boardError) {
+			return new DataResponse($boardError['message'], $boardError['status']);
+		}
+
+		$stackError = $this->apiHelper->entityHasError($this->request->params['stackId'], 'stack', $this->stackService);		
+		if ($stackError) {
+			return new DataResponse($stackError['message'], $stackError['status']);
+		}
+
+		if (is_numeric($this->request->params['cardId']) === false) {
+			return new DataResponse('card id must be a number', HTTP::STATUS_BAD_REQUEST);
+		}
+
+		$card = $this->cardService->find($this->request->params['cardId']);
```

Comment:
Something more general regarding the service classes. They actually throw various exceptions that come from either the permissionService or the db mappers. Sorry those are not correctly documented in the phpDoc comments. So we need to either add a try/catch block here or extend the service methods to return null in case there is no enity found.

Another approach would be to have a middleware for the API that catches the exceptions and returns a 404 for DoesNotExistException for example.

Reply:
Which would you prefer? I'm usually all for the try { catch } blocks as it makes the code that much more clear / self documenting, typically avoiding exactly this kind of scenario. I did a test for (false || null) just as a starting point.

All the services are able to throw these exceptions?

Reply:
Yes, that makes sense. Let's go that way. Basically yes, but I can push some commit to add proper annotations later, if that helps.